libraries/GCS_MAVLink/GCS_FTP.cpp

```diff
@@ -0,0 +1,604 @@
+
+                        // must actually be reading a file
+                        if (ftp.fd == -1) {
+                            ftp_error(reply, FTP_ERROR::FileNotFound);
+                            break;
+                        }
+
+                        // must have the file in read mode
+                        if ((ftp.mode != FTP_FILE_MODE::Write)) {
+                            ftp_error(reply, FTP_ERROR::Fail);
+                            break;
+                        }
+
+                        // seek to requested offset
+                        if (AP::FS().lseek(ftp.fd, request.offset, SEEK_SET) == -1) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+
+                        // fill the buffer
+                        ssize_t write_bytes = AP::FS().write(ftp.fd, request.data, request.size);
+                        if (write_bytes == -1) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+
+                        reply.opcode = FTP_OP::Ack;
+                        reply.offset = request.offset;
+                        break;
+                    }
+                case FTP_OP::CreateDirectory:
+                    {
+                        // sanity check that our the request looks well formed
+                        const size_t file_name_len = strnlen((char *)request.data, sizeof(request.data));
+                        if ((file_name_len != request.size) || (request.size == 0)) {
+                            ftp_error(reply, FTP_ERROR::InvalidDataSize);
+                            break;
+                        }
+
+                        request.data[sizeof(request.data) - 1] = 0; // ensure the path is null terminated
+
+                        // actually open the file
+                        if (AP::FS().mkdir((char *)request.data) == -1) {
+                            ftp_error(reply, FTP_ERROR::FailErrno);
+                            break;
+                        }
+
+                        reply.opcode = FTP_OP::Ack;
+                        break;
+                    }
+                case FTP_OP::RemoveDirectory:
+                case FTP_OP::RemoveFile:
+                    {
+                        // sanity check that our the request looks well formed
+                        const size_t file_name_len = strnlen((char *)request.data, sizeof(request.data));
+                        if ((file_name_len != request.size) || (request.size == 0)) {
+                            ftp_error(reply, FTP_ERROR::InvalidDataSize);
+                            break;
+                        }
+
+                        request.data[sizeof(request.data) - 1] = 0; // ensure the path is null terminated
+
+                        // actually open the file
+                        if (AP::FS().unlink((char *)request.data) == -1) {
```

Comment:
Has this been tested under AP_HAL_Linux?

Reply:
HAL_SITL, which uses the same filesystem posix stuff to the best of my knowledge. Bad comment on the line above though :)